llarp/CMakeLists.txt

```
@@ -242,65 +306,218 @@ add_library(lokinet-amalgum
   service/tag.cpp
 )
 
-set(BOOTSTRAP_FALLBACKS)
-foreach(bs IN ITEMS MAINNET TESTNET)
-  if(BOOTSTRAP_FALLBACK_${bs})
-    message(STATUS "Building with ${bs} fallback boostrap path \"${BOOTSTRAP_FALLBACK_${bs}}\"")
-    file(READ "${BOOTSTRAP_FALLBACK_${bs}}" bs_data HEX)
-    if(bs STREQUAL TESTNET)
-      set(network "gamma")
-    elseif(bs STREQUAL MAINNET)
-      set(network "lokinet")
-    else()
-      string(TOLOWER "${bs}" network)
-    endif()
-    string(REGEX REPLACE "([0-9a-f][0-9a-f])" "\\\\x\\1" bs_data "${bs_data}")
-    set(BOOTSTRAP_FALLBACKS "${BOOTSTRAP_FALLBACKS}{\"${network}\"s, \"${bs_data}\"sv},\n")
-  endif()
-endforeach()
-configure_file("bootstrap-fallbacks.cpp.in" "${CMAKE_CURRENT_BINARY_DIR}/bootstrap-fallbacks.cpp" @ONLY)
-target_sources(lokinet-amalgum PRIVATE "${CMAKE_CURRENT_BINARY_DIR}/bootstrap-fallbacks.cpp")
+add_library(lokinet-layer-platform
+  STATIC
+  layers/platform/stub.cpp # todo: remove me
+)
 
-if(WITH_PEERSTATS_BACKEND)
-  target_compile_definitions(lokinet-amalgum PRIVATE -DLOKINET_PEERSTATS_BACKEND)
-  target_link_libraries(lokinet-amalgum PUBLIC sqlite_orm)
-endif()
 
+# interal tooling for pybind
+add_library(lokinet-tooling INTERFACE)
 if(WITH_HIVE)
-  target_sources(lokinet-amalgum PRIVATE
+  add_library(lokinet-hive-tooling
+    STATIC
     tooling/router_hive.cpp
     tooling/hive_router.cpp
     tooling/hive_context.cpp
   )
+  target_link_libraries(lokinet-tooling INTERFACE lokinet-hive-tooling)
+endif()
+
+
+# interface library for setting commone includes, linkage and flags.
+add_library(lokinet-base INTERFACE)
+target_include_directories(lokinet-base
+  INTERFACE ${PROJECT_SOURCE_DIR} ${PROJECT_SOURCE_DIR}/include
+)
+target_link_libraries(lokinet-base INTERFACE oxen::logging lokinet-cryptography)
+
+if(WITH_PEERSTATS)
+  target_compile_definitions(lokinet-base INTERFACE -DLOKINET_PEERSTATS_BACKEND)
+  target_link_libraries(lokinet-base INTERFACE sqlite_orm)
 endif()
 
-# TODO: make libunbound hidden behind a feature flag like sqlite for embedded lokinet
-target_link_libraries(lokinet-amalgum PRIVATE libunbound)
+# interface libraries for internal linkage
+add_library(lokinet-layers INTERFACE)
+add_library(lokinet-amalgum INTERFACE)
+
+
+# helper function to link a library to lokinet-base, enable lto, add to lokinet-amalgum and then link to other libs
+function(lokinet_link_lib)
+  set(libname ${ARGV0})
+  message(DEBUG "created target: ${libname}")
+  list(POP_FRONT ARGV)
+  enable_lto(${libname})
+  target_link_libraries(${libname} PUBLIC lokinet-base ${ARGV})
+  target_link_libraries(lokinet-amalgum INTERFACE ${libname})
+endfunction()
```

CMake functions can have arguments which would save the ARGV manipulation here; any argument passed beyond the listed ones then go into ${ARGN}.  So this would simplify a bit to:
```suggestion
function(lokinet_link_lib libname)
  message(DEBUG "created target: ${libname}")
  enable_lto(${libname})
  target_link_libraries(${libname} PUBLIC lokinet-base ${ARGN})
  target_link_libraries(lokinet-amalgum INTERFACE ${libname})
endfunction()
```